op-supervisor/supervisor/backend/rewinder/rewinder.go

```diff
@@ -0,0 +1,239 @@
+			return fmt.Errorf("failed to rewind local for chain %s: %w", chainID, err)
+		}
+	}
+
+	return nil
+}
+
+// attemptRewindUnsafe attempts to rewind unsafe blocks
+func (r *Rewinder) attemptRewindUnsafe(chainID eth.ChainID, badBlockHeight uint64) error {
+	return r.attemptRewind(chainID, badBlockHeight, rewindController{
+		isSafe:       false,
+		getFinalized: derivedFromPairGetter(r.db.LocalSafe),
+		getLocal:     r.db.LocalUnsafe,
+		getCross:     r.db.CrossUnsafe,
+		rewindLocal:  r.db.RewindLocalUnsafe,
+		rewindCross:  r.db.RewindCrossUnsafe,
+	})
+}
+
+// attemptRewindSafe attempts to rewind safe blocks
+func (r *Rewinder) attemptRewindSafe(chainID eth.ChainID, badBlockHeight uint64) error {
+	return r.attemptRewind(chainID, badBlockHeight, rewindController{
+		isSafe:       true,
+		getFinalized: r.db.Finalized,
+		getLocal:     derivedFromPairGetter(r.db.LocalSafe),
+		getCross:     derivedFromPairGetter(r.db.CrossSafe),
+		rewindLocal:  r.db.RewindLocalSafe,
+		rewindCross:  r.db.RewindCrossSafe,
+	})
+}
+
+// findLatestCommonAncestor finds the latest common ancestor between the startBlock and the finalizedBlock
+// by searching for the last block that exists in both the local db and the L2 node.
+// If no common ancestor is found then the finalized head is returned.
+func (r *Rewinder) findLatestCommonAncestor(chainID eth.ChainID, startBlock uint64, finalizedBlock types.BlockSeal) (types.BlockSeal, error) {
+	syncNode, ok := r.syncNodes.Get(chainID)
+	if !ok {
+		return types.BlockSeal{}, fmt.Errorf("sync node not found for chain %s", chainID)
+	}
+
+	// Linear search from local height down to finalized height
+	r.log.Info("searching for common ancestor", "chain", chainID, "local", startBlock, "finalized", finalizedBlock.Number)
+	finalizedHeight := int64(finalizedBlock.Number)
+	for height := int64(startBlock); height >= finalizedHeight; height-- {
+		// Load the block at this height from the node and the local db
+		remoteRef, err := syncNode.BlockRefByNumber(context.Background(), uint64(height))
+		if err != nil {
+			return types.BlockSeal{}, err
+		}
+		localRef, err := r.db.FindSealedBlock(chainID, uint64(height))
+		if err != nil {
+			return types.BlockSeal{}, err
+		}
+
+		// If the block is the same then we've found the common ancestor
+		if localRef.Hash == remoteRef.Hash {
+			return localRef, nil
+		}
+	}
+
+	// If we didn't find a common ancestor then return the finalized head
+	return finalizedBlock, nil
+}
+
+// derivedFromPairGetter wraps a (chainID) -> (types.DerivedBlockSealPair) function
+// and returns the derived block seal from the pair instead of the pair itself.
+func derivedFromPairGetter(fn func(eth.ChainID) (types.DerivedBlockSealPair, error)) func(eth.ChainID) (types.BlockSeal, error) {
+	return func(chainID eth.ChainID) (types.BlockSeal, error) {
+		pair, err := fn(chainID)
+		if err != nil {
+			return types.BlockSeal{}, err
+		}
+		return pair.Derived, nil
+	}
+}
```

Comment:
Rather than wrap/transform the getters being used in the rewind controller, why not just have the rewind controller use pairs?

Reply:
I've tried to make this a _little_ bit clearer in the doc string. I do agree it looks a little weird but I think it's probably the cleanest way to do it.